### PR TITLE
Fixing circular padding in nn.functional.py

### LINKED ISSUE #24504: Circular padding doesn't work when the 'right' side is 0
Reproduction:
```
import torch
import torch.nn.functional as F

inputs = torch.randn(1, 1, 5, 5)
print(F.pad(inputs, [4,0,3,0], mode='circular').shape) # (1,1,5,5) (Wrong)
print(F.pad(inputs, [1,0,0,0], mode='circular').shape) # (1,1,5,5) (Wrong)
print(F.pad(inputs, [1,1,0,0], mode='circular').shape) # (1,1,5,7) (Correct)
print(F.pad(inputs, [2,1,0,0], mode='circular').shape) # (1,1,5,8) (Correct)
```

Correct results:
```
import torch
import torch.nn.functional as F

inputs = torch.randn(1, 1, 5, 5)
print(F.pad(inputs, [4,0,3,0], mode='circular').shape) # (1,1,8,9)
print(F.pad(inputs, [1,0,0,0], mode='circular').shape) # (1,1,5,6)
print(F.pad(inputs, [1,1,0,0], mode='circular').shape) # (1,1,5,7)
print(F.pad(inputs, [2,1,0,0], mode='circular').shape) # (1,1,5,8)
```

cc @SsnL

### PR DESCRIPTION
**Pull request by pdlive215**
Fixes #24504. Note that `_pad_circular` implements padding in Python, whereas other padding implementations are in `torch/csrc`, and defined in `nn.modules.padding`. Thus this is a bit of a "workaround of a workaround," and I think we should implement circular padding in `torch/csrc` and `nn.modules.padding` going forward.

### CODE REVIEW

**Comment by Chillee** [comment]
Could you add a test for this case?

**Comment by pdlive215** [comment]
and that's why you don't use vi as an IDE...

Is there a guide for writing test cases? I don't see anything in the contribution guide other than "additional test coverage is appreciated."

**Comment by Chillee** [comment]
I don't think so. I usually just look for tests that touch code near what I modified and figure it out from there. For example, I would search for 'circular' in the `test/` directory to find the relevant tests here. You can also ask in the PyTorch slack.

I would put these tests in particular here: https://github.com/pytorch/pytorch/blob/master/test/common_nn.py#L2336

I think in this case you might need to change the `reference_fn` `padding1d_circular` since that's what you're fixing in the first place.

**Comment by pdlive215** [comment]
Cool, so I can think of each test (`padding1d_circular`, `padding2d_circular`, and `padding3d_circular`) as specifications of `_pad_circular`, and just rewrite `_pad_circular` for each instance?

FYI the existing `padding1d_circular` test is failing with a `TypeError` for the sample input and padding. Will make sure to fix this.

File: test/common_nn.py [hunk 1]
```diff
@@ -2679,74 +2679,95 @@ def ctcloss_reference(log_probs, targets, input_lengths, target_lengths, blank=0
...
 
 def padding1d_circular(input, pad):
-    r""" input:
-            [[[0., 1., 2.],
-              [3., 4., 5.]]]
-          pad: (1, 2)
-          output:
-            [[[2., 0., 1., 2., 0., 1.],
-              [5., 3., 4., 5., 3., 4.]]]
+    r"""input: torch.Tensor([[[0., 1., 2.],
```
Comment (gchanan) [inline test/common_nn.py]: don't use the old-style constructors, use "torch.tensor" instead.
